go/enclave/components/interfaces.go

```diff
@@ -0,0 +1,138 @@
+}
+
+type BatchRegistry interface {
+	// StoreBatch - will store the batch and receipts in storage.
+	// Furthermore any heads and pointers would be updated here and
+	// after all is done the batch will be pushed to the subscribers
+	// in order to update them.
+	StoreBatch(*core.Batch, types.Receipts) error
+
+	// GetHeadBatch - Returns the batch considered to be the L2 head.
+	GetHeadBatch() (*core.Batch, error)
+
+	// GetHeadBatchFor - Returns the head batch for given L1 block.
+	// Note that this does not seek backwards so if a given block does not have
+	// any pointers to batches then `ErrNotFound` will be returned.
+	GetHeadBatchFor(common.L1BlockHash) (*core.Batch, error)
+
+	// GetBatch - Given a batch hash returns the matching batch taken from the
+	// storage. If no such batch is found then nil and `ErrNotFound` should
+	// be returned.
+	GetBatch(common.L2BatchHash) (*core.Batch, error)
+
+	// FindAncestralBatchFor - this function is akin to `GetHeadBatchFor`, but
+	// it does not check the block passed as a parameter, but rather starts from its
+	// parent block and then seeks backwards until a head batch is found or genesis is reached.
+	FindAncestralBatchFor(*common.L1Block) (*core.Batch, error)
+
+	// BatchesAfter - Given a hash, will return batches following it until the head batch
+	BatchesAfter(batchHash gethcommon.Hash) ([]*core.Batch, error)
+
+	// GetBatchStateAtHeight - creates a stateDB that represents the state committed when
+	// the batch with height matching the blockNumber was created and stored.
+	GetBatchStateAtHeight(blockNumber *gethrpc.BlockNumber) (*state.StateDB, error)
+
+	// GetBatchStateAtHeight - same as `GetBatchStateAtHeight`, but instead returns the full batch
+	// rather than its stateDB only.
+	GetBatchAtHeight(height gethrpc.BlockNumber) (*core.Batch, error)
+
+	// Subscribe - creates and returns a channel that will be used to push any newly created batches
+	// to the subscriber.
+	Subscribe() chan *core.Batch
+	// Unsubscribe - informs the registry that the subscriber is no longer listening, allowing it to
+	// gracefully terminate any streaming and stop queueing new batches.
+	Unsubscribe()
+
+	// HasGenesisBatch - returns if genesis batch is available yet or not, or error in case
+	// the function is unable to determine.
+	HasGenesisBatch() (bool, error)
+}
+
+type RollupProducer interface {
+	// CreateRollup - creates a rollup starting from the end of the last rollup
+	// that has been stored and continues it towards what we consider the current L2 head.
+	CreateRollup() (*core.Rollup, error)
+}
+
+type RollupConsumer interface {
+	// ProcessL1Block - extracts the rollups from the block's transactions
+	// and verifies their integrity, saving and processing any batches that have
+	// not been seenp previously.
+	ProcessL1Block(b *common.BlockAndReceipts) ([]*core.Rollup, error)
+}
+
+// TODO - find a better place for the following stuff
```

Comment:
why not move them to the batch_producer

Reply:
wasn't clear at the time what the best place would be